walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    Collection<TimeRange> answer = new ArrayList<TimeRange>();
+
+    Collection<String> requestAttendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+    
+    //Filter events to only consider events that have attendees also in the Meeting Requested
+    List<TimeRange> relevantEventTimes = filterEvents(events, requestAttendees);
+    
+    Collections.sort(relevantEventTimes, TimeRange.ORDER_BY_START);
+
+    //Variables used to resolve time range overlaps during linear scan 
+    int combinedRangeStartTime;
+    int combinedRangeEndTime = TimeRange.START_OF_DAY;    
+  
+    //Linear scan to assess all possible time slots between and before events
+    for (TimeRange currentTimeRange : relevantEventTimes) {
+      int currentStartTime = currentTimeRange.start();
+      int currentEndTime = currentTimeRange.end();
+      
+      if (currentStartTime <= combinedRangeEndTime) {
+        if (currentEndTime > combinedRangeEndTime) {
+          combinedRangeEndTime = currentEndTime;
+        }
+      } else {
+        addPossibleTimeRange(answer, combinedRangeEndTime, currentStartTime, requestDuration, false);
+        combinedRangeStartTime = currentStartTime;
+        combinedRangeEndTime = currentEndTime;
+      }
+    }
+    
+    //Process the timeRange from end of last event to end of day, or the entire day if there are no events
+    addPossibleTimeRange(answer, combinedRangeEndTime, TimeRange.END_OF_DAY, requestDuration, true);
+    
+    return answer;
+  }
+  
+ 
+  /**
+   * Helper function to filter through the events given and only return the time ranges 
+   * for events that have attendees in the MeetingRequest's attendees list.
+   * 
+   * @param events
+   * @param requestAttendees
+   * @return List of TimeRange for events that have attendees who are also in the requested meeting
+   */
+  private List<TimeRange> filterEvents(Collection<Event> events, Collection<String> requestAttendees) {
+    Iterator<Event> eventsIterator = events.iterator(); 
+    List<TimeRange> relevantEventTimes = new ArrayList<TimeRange>();
+    while (eventsIterator.hasNext()) {
```

Comment:
Prefer short-hand over iterator.

for (Event event : events) {...}

Reply:
fixed!